### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) for cargo-zigbuild and cargo-xwin

### DISCUSSION THREAD

**Discussion by zamazan4ik** [rust-cross/cargo-zigbuild#280] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-binstall` Release [profile](https://github.com/cargo-bins/cargo-binstall/blob/main/Cargo.toml#L21).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I made quick tests locally (Rustc 1.81, latest project versions, Fedora 40):
* `cargo-zigbuild` binary size reduction from enabling `lto = true`: from 4.1 Mib to 3.2 Mib
* `cargo-xwin` binary size reduction from enabling `lto = true`: from 12 Mib to 8.4 Mib

Thank you.

**Comment by polarathene** [comment]
> I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

It is better when you have metrics that can convey such differences as valuable. On a developer machine saving 1MB or so won't really matter in comparison to the full toolchain required. Usually you want to minimize on size for distribution of actual projects, similar for performance you need to measure.

Thin LTO is usually convenient though. In my experience quite a few projects that adopt `lto = true` on the hand become much more annoying to build/install (eg `cargo install crate-name`) where their LTO enablement raises required RAM to build their project considerably, thus one needs to do workarounds. `lto = true` is better as a custom release profile such as `dist` for CI to build when publishing actual release artifacts.
